docs/internal/interop.md

```diff
@@ -0,0 +1,351 @@
+   where calls to external functions go one way, and calls to allocation, system
+   calls, etc. go the other way.
+6. The appropriate object files are created and linked with the main executable.
+
+Not all functions will be representable in Verona, not all structure layouts will
+be easily known at the Verona level (special hardware ABI), so errors may still
+occur here.
+
+The compiler is responsible for emitting errors that inform the programmer what
+type/function cannot be represented and potentially why. The programmer is
+responsible for implementing a wrapper function that abstracts away the
+complexity of the language in a public function that Verona can call.
+
+Language support will be always evolving. This is expected to be a balance
+between "best effort" and "appropriate support".
+
+### Run time
+
+1. When the sandbox is initialised, its allocator sets up a new isolated heap
+   that is used for all allocations in the sandbox (including Verona foreign
+   objects that belong to that region).
+2. When the foreign code makes calls to allocate memory or run potentially unsafe
+   operations (file descriptors, system calls, etc.), the RPC layer will check
+   them and forward if explicitly marked as an upcall or if it's a direct
+   sandbox functionality (webasm, CHERI), otherwise fail and close the sandbox.
+3. When Verona code makes calls to the foreign library, the compiler-generated
+   serialisation code will transform the arguments and return value into a
+   memory buffer and call the dispatcher with the appropriate index and buffer.
+4. The dispatcher will call a templated wrapper that unpacks the arguments and
+   calls the actual function.
+5. The return value, if any, is desserialised and bound to the Verona variable
+   or used as an argument in a Verona call.
+
+## C/C++ interoperability
+
+For both C and C++, we use `clang` to parse the header file to know what types
+are declared, functions are exported, etc. This allows us to know which of the
+Verona code lines calling foreign code are correct, and generate the correct
+serialisation routines and, for C++, instantiating the correct template
+implementations based on the types used in Verona.
+
+The C++ interoperability layer parses the file and generates the foreign AST.
+We then use the Verona code function names and types to create
+a function declaration (context, name, argument types) and, if valid, we add
+that function to the dispatcher, giving it its own (auto-incremented) index.
+
+Additional code needs to be generated to serialise the Verona arguments and
+return values into a buffer, so that the dispatcher can be generic, only taking
+an index and a memory buffer.
+
+For C++, further code will be created (via AST construction and LLVM code
+generation) if the types involved are parametrised. All template arguments are
+taken into consideration and a specialisation is defined and generated in a
+wrapper layer.
```

Comment:
Probably worth noting that, at least for v1, we don't allow Verona generic parameters to be passed to template instantiations.  This means that all possible template instantiations that a Verona generic may use must exist *before* reification, which means we have the same guarantees for Verona generics that use C++ templates as we do for other Verona generics (i.e. they either type check and work for all instantiations or they don't type check).

We may relax that later, but I believe we should probably do that only once C++ Concepts have landed and then we can map concepts into Verona interface types and keep the same guarantees.